File: ts-sdk/src/cel/parser.ts
/**
 * CEL Parser - parse tokens into AST
 */

import { Lexer, Token, TokenType } from './lexer.js';

export type ASTNode =
  | LiteralNode
  | IdentifierNode
  | BinaryNode
  | UnaryNode
  | MemberNode
  | IndexNode
  | CallNode
  | TernaryNode
  | ListNode
  | MapNode;

export interface LiteralNode {
  type: 'Literal';
  value: string | number | boolean | null;
}

export interface IdentifierNode {
  type: 'Identifier';
  name: string;
}

export interface BinaryNode {
  type: 'Binary';
  operator: string;
  left: ASTNode;
  right: ASTNode;
}

export interface UnaryNode {
  type: 'Unary';
  operator: string;
  operand: ASTNode;
}

export interface MemberNode {
  type: 'Member';
  object: ASTNode;
  property: string;
}

export interface IndexNode {
  type: 'Index';
  object: ASTNode;
  index: ASTNode;
}

export interface CallNode {
  type: 'Call';
  callee: ASTNode;
  args: ASTNode[];
}

export interface TernaryNode {
  type: 'Ternary';
  condition: ASTNode;
  consequent: ASTNode;
  alternate: ASTNode;
}

export interface ListNode {
  type: 'List';
  elements: ASTNode[];
}

export interface MapNode {
  type: 'Map';
  entries: Array<{ key: ASTNode; value: ASTNode }>;
}

export class Parser {
  private tokens: Token[] = [];
  private pos = 0;

  parse(input: string): ASTNode {
    const lexer = new Lexer(input);
    this.tokens = lexer.tokenize();
    this.pos = 0;

    const result = this.parseExpression();

    if (this.current().type !== 'EOF') {
      throw new Error(`Unexpected token '${this.current().value}' at position ${this.current().pos}`);
    }

    return result;
  }

  private current(): Token {
    return this.tokens[this.pos] ?? { type: 'EOF', value: null, pos: -1 };
  }

  private advance(): Token {
    const token = this.current();
    this.pos++;
    return token;
  }

  private expect(type: TokenType): Token {
    const token = this.current();
    if (token.type !== type) {
      throw new Error(`Expected ${type} but got ${token.type} at position ${token.pos}`);
    }
    return this.advance();
  }

  private parseExpression(): ASTNode {
    return this.parseTernary();
  }

  private parseTernary(): ASTNode {
    let condition = this.parseOr();

    if (this.current().type === 'QUESTION') {
      this.advance(); // consume ?
      const consequent = this.parseExpression();
      this.expect('COLON');
      const alternate = this.parseExpression();
      condition = {
        type: 'Ternary',
        condition,
        consequent,
        alternate,
      };
    }

    return condition;
  }

  private parseOr(): ASTNode {
    let left = this.parseAnd();

    while (this.current().type === 'OR') {
      const operator = this.advance().value as string;
      const right = this.parseAnd();
      left = { type: 'Binary', operator, left, right };
    }

    return left;
  }

  private parseAnd(): ASTNode {
    let left = this.parseEquality();

    while (this.current().type === 'AND') {
      const operator = this.advance().value as string;
      const right = this.parseEquality();
      left = { type: 'Binary', operator, left, right };
    }

    return left;
  }

  private parseEquality(): ASTNode {
    let left = this.parseComparison();

    while (this.current().type === 'EQ' || this.current().type === 'NEQ') {
      const operator = this.advance().value as string;
      const right = this.parseComparison();
      left = { type: 'Binary', operator, left, right };
    }

    return left;
  }

  private parseComparison(): ASTNode {
    let left = this.parseIn();

    while (
      this.current().type === 'LT' ||
      this.current().type === 'LTE' ||
      this.current().type === 'GT' ||
      this.current().type === 'GTE'
    ) {
      const operator = this.advance().value as string;
      const right = this.parseIn();
      left = { type: 'Binary', operator, left, right };
    }

    return left;
  }

  private parseIn(): ASTNode {
    let left = this.parseAdditive();

    if (this.current().type === 'IN') {
      this.advance(); // consume 'in'
      const right = this.parseAdditive();
      left = { type: 'Binary', operator: 'in', left, right };
    }

    return left;
  }

  private parseAdditive(): ASTNode {
    let left = this.parseMultiplicative();

    while (this.current().type === 'PLUS' || this.current().type === 'MINUS') {
      const operator = this.advance().value as string;
      const right = this.parseMultiplicative();
      left = { type: 'Binary', operator, left, right };
    }

    return left;
  }

  private parseMultiplicative(): ASTNode {
    let left = this.parseUnary();

    while (
      this.current().type === 'STAR' ||
      this.current().type === 'SLASH' ||
      this.current().type === 'PERCENT'
    ) {
      const operator = this.advance().value as string;
      const right = this.parseUnary();
      left = { type: 'Binary', operator, left, right };
    }

    return left;
  }

  private parseUnary(): ASTNode {
    if (this.current().type === 'NOT' || this.current().type === 'MINUS') {
      const operator = this.advance().value as string;
      const operand = this.parseUnary();
      return { type: 'Unary', operator, operand };
    }

    return this.parsePostfix();
  }

  private parsePostfix(): ASTNode {
    let node = this.parsePrimary();

    while (true) {
      if (this.current().type === 'DOT') {
        this.advance(); // consume .
        const property = this.expect('IDENT').value as string;
        node = { type: 'Member', object: node, property };
      } else if (this.current().type === 'LBRACKET') {
        this.advance(); // consume [
        const index = this.parseExpression();
        this.expect('RBRACKET');
        node = { type: 'Index', object: node, index };
      } else if (this.current().type === 'LPAREN') {
        this.advance(); // consume (
        const args: ASTNode[] = [];
        if (this.current().type !== 'RPAREN') {
          args.push(this.parseExpression());
          while (this.current().type === 'COMMA') {
            this.advance(); // consume ,
            args.push(this.parseExpression());
          }
        }
        this.expect('RPAREN');
        node = { type: 'Call', callee: node, args };
      } else {
        break;
      }
    }

    return node;
  }

  private parsePrimary(): ASTNode {
    const token = this.current();

    switch (token.type) {
      case 'NUMBER':
      case 'STRING':
      case 'BOOL':
      case 'NULL':
        this.advance();
        return { type: 'Literal', value: token.value };

      case 'IDENT':
        this.advance();
        return { type: 'Identifier', name: token.value as string };

      case 'LPAREN': {
        this.advance(); // consume (
        const expr = this.parseExpression();
        this.expect('RPAREN');
        return expr;
      }

      case 'LBRACKET': {
        this.advance(); // consume [
        const elements: ASTNode[] = [];
        if (this.current().type !== 'RBRACKET') {
          elements.push(this.parseExpression());
          while (this.current().type === 'COMMA') {
            this.advance(); // consume ,
            elements.push(this.parseExpression());
          }
        }
        this.expect('RBRACKET');
        return { type: 'List', elements };
      }

      default:
        throw new Error(`Unexpected token '${token.type}' at position ${token.pos}`);
    }
  }
}

File: rust/ais-rs/crates/ais-cel/src/parser.rs
use crate::ast::{AstNode, BinaryOp, UnaryOp};
use crate::lexer::{tokenize, LexError, Token, TokenKind};

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum ParseError {
    #[error("lex error: {0}")]
    Lex(#[from] LexError),
    #[error("unexpected token at {pos}: expected {expected}, got {found}")]
    UnexpectedToken {
        expected: String,
        found: String,
        pos: usize,
    },
    #[error("invalid number literal at {pos}: {literal}")]
    InvalidNumber { literal: String, pos: usize },
}

pub struct Parser {
    tokens: Vec<Token>,
    index: usize,
}

pub fn parse_expression(input: &str) -> Result<AstNode, ParseError> {
    let tokens = tokenize(input)?;
    Parser::new(tokens).parse()
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, index: 0 }
    }

    pub fn parse(mut self) -> Result<AstNode, ParseError> {
        let expression = self.parse_ternary()?;
        self.expect(TokenKind::Eof)?;
        Ok(expression)
    }

    fn parse_ternary(&mut self) -> Result<AstNode, ParseError> {
        let mut condition = self.parse_or()?;
        if self.match_kind(TokenKind::Question) {
            let then_expr = self.parse_ternary()?;
            self.expect(TokenKind::Colon)?;
            let else_expr = self.parse_ternary()?;
            condition = AstNode::Ternary {
                condition: Box::new(condition),
                then_expr: Box::new(then_expr),
                else_expr: Box::new(else_expr),
            };
        }
        Ok(condition)
    }

    fn parse_or(&mut self) -> Result<AstNode, ParseError> {
        self.parse_binary_chain(|token| matches!(token, TokenKind::OrOr), Self::parse_and, BinaryOp::Or)
    }

    fn parse_and(&mut self) -> Result<AstNode, ParseError> {
        self.parse_binary_chain(|token| matches!(token, TokenKind::AndAnd), Self::parse_equality, BinaryOp::And)
    }

    fn parse_equality(&mut self) -> Result<AstNode, ParseError> {
        let mut node = self.parse_comparison()?;
        loop {
            if self.match_kind(TokenKind::EqEq) {
                let right = self.parse_comparison()?;
                node = AstNode::Binary {
                    left: Box::new(node),
                    op: BinaryOp::Eq,
                    right: Box::new(right),
                };
                continue;
            }
            if self.match_kind(TokenKind::NotEq) {
                let right = self.parse_comparison()?;
                node = AstNode::Binary {
                    left: Box::new(node),
                    op: BinaryOp::Ne,
                    right: Box::new(right),
                };
                continue;
            }
            break;
        }
        Ok(node)
    }

    fn parse_comparison(&mut self) -> Result<AstNode, ParseError> {
        let mut node = self.parse_term()?;
        loop {
            let op = if self.match_kind(TokenKind::Lt) {
                Some(BinaryOp::Lt)
            } else if self.match_kind(TokenKind::Le) {
                Some(BinaryOp::Le)
            } else if self.match_kind(TokenKind::Gt) {
                Some(BinaryOp::Gt)
            } else if self.match_kind(TokenKind::Ge) {
                Some(BinaryOp::Ge)
            } else if self.match_kind(TokenKind::In) {
                Some(BinaryOp::In)
            } else {
                None
            };

            let Some(op) = op else { break };
            let right = self.parse_term()?;
            node = AstNode::Binary {
                left: Box::new(node),
                op,
                right: Box::new(right),
            };
        }
        Ok(node)
    }

    fn parse_term(&mut self) -> Result<AstNode, ParseError> {
        let mut node = self.parse_factor()?;
        loop {
            let op = if self.match_kind(TokenKind::Plus) {
                Some(BinaryOp::Add)
            } else if self.match_kind(TokenKind::Minus) {
                Some(BinaryOp::Sub)
            } else {
                None
            };
            let Some(op) = op else { break };
            let right = self.parse_factor()?;
            node = AstNode::Binary {
                left: Box::new(node),
                op,
                right: Box::new(right),
            };
        }
        Ok(node)
    }

    fn parse_factor(&mut self) -> Result<AstNode, ParseError> {
        let mut node = self.parse_unary()?;
        loop {
            let op = if self.match_kind(TokenKind::Star) {
                Some(BinaryOp::Mul)
            } else if self.match_kind(TokenKind::Slash) {
                Some(BinaryOp::Div)
            } else if self.match_kind(TokenKind::Percent) {
                Some(BinaryOp::Mod)
            } else {
                None
            };
            let Some(op) = op else { break };
            let right = self.parse_unary()?;
            node = AstNode::Binary {
                left: Box::new(node),
                op,
                right: Box::new(right),
            };
        }
        Ok(node)
    }

    fn parse_unary(&mut self) -> Result<AstNode, ParseError> {
        if self.match_kind(TokenKind::Bang) {
            let expr = self.parse_unary()?;
            return Ok(AstNode::Unary {
                op: UnaryOp::Not,
                expr: Box::new(expr),
            });
        }
        if self.match_kind(TokenKind::Minus) {
            let expr = self.parse_unary()?;
            return Ok(AstNode::Unary {
                op: UnaryOp::Neg,
                expr: Box::new(expr),
            });
        }
        self.parse_postfix()
    }

    fn parse_postfix(&mut self) -> Result<AstNode, ParseError> {
        let mut node = self.parse_primary()?;

        loop {
            if self.match_kind(TokenKind::Dot) {
                let token = self.expect(TokenKind::Identifier)?;
                node = AstNode::Member {
                    object: Box::new(node),
                    property: token.lexeme,
                };
                continue;
            }

            if self.match_kind(TokenKind::LBracket) {
                let index = self.parse_ternary()?;
                self.expect(TokenKind::RBracket)?;
                node = AstNode::Index {
                    object: Box::new(node),
                    index: Box::new(index),
                };
                continue;
            }

            if self.match_kind(TokenKind::LParen) {
                let mut args = Vec::new();
                if !self.check(TokenKind::RParen) {
                    loop {
                        args.push(self.parse_ternary()?);
                        if self.match_kind(TokenKind::Comma) {
                            continue;
                        }
                        break;
                    }
                }
                self.expect(TokenKind::RParen)?;
                node = AstNode::Call {
                    callee: Box::new(node),
                    args,
                };
                continue;
            }

            break;
        }

        Ok(node)
    }

    fn parse_primary(&mut self) -> Result<AstNode, ParseError> {
        let token = self.advance();
        match token.kind {
            TokenKind::Identifier => Ok(AstNode::Identifier(token.lexeme)),
            TokenKind::Integer => Ok(AstNode::Integer(token.lexeme)),
            TokenKind::Decimal => Ok(AstNode::Decimal(token.lexeme)),
            TokenKind::String => Ok(AstNode::String(token.lexeme)),
            TokenKind::True => Ok(AstNode::Bool(true)),
            TokenKind::False => Ok(AstNode::Bool(false)),
            TokenKind::Null => Ok(AstNode::Null),
            TokenKind::LParen => {
                let expr = self.parse_ternary()?;
                self.expect(TokenKind::RParen)?;
                Ok(expr)
            }
            TokenKind::LBracket => {
                let mut items = Vec::new();
                if !self.check(TokenKind::RBracket) {
                    loop {
                        items.push(self.parse_ternary()?);
                        if self.match_kind(TokenKind::Comma) {
                            continue;
                        }
                        break;
                    }
                }
                self.expect(TokenKind::RBracket)?;
                Ok(AstNode::List(items))
            }
            _ => Err(ParseError::UnexpectedToken {
                expected: "primary expression".to_string(),
                found: format!("{:?}", token.kind),
                pos: token.pos,
            }),
        }
    }

    fn parse_binary_chain<F, G>(
        &mut self,
        matcher: F,
        mut parse_operand: G,
        op: BinaryOp,
    ) -> Result<AstNode, ParseError>
    where
        F: Fn(&TokenKind) -> bool,
        G: FnMut(&mut Self) -> Result<AstNode, ParseError>,
    {
        let mut node = parse_operand(self)?;
        while matcher(&self.peek().kind) {
            self.advance();
            let right = parse_operand(self)?;
            node = AstNode::Binary {
                left: Box::new(node),
                op,
                right: Box::new(right),
            };
        }
        Ok(node)
    }

    fn check(&self, kind: TokenKind) -> bool {
        self.peek().kind == kind
    }

    fn match_kind(&mut self, kind: TokenKind) -> bool {
        if self.check(kind) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token, ParseError> {
        let token = self.advance();
        if token.kind == kind {
            Ok(token)
        } else {
            Err(ParseError::UnexpectedToken {
                expected: format!("{:?}", kind),
                found: format!("{:?}", token.kind),
                pos: token.pos,
            })
        }
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.index]
    }

    fn advance(&mut self) -> Token {
        let token = self.tokens[self.index].clone();
        if self.index + 1 < self.tokens.len() {
            self.index += 1;
        }
        token
    }
}

#[cfg(test)]
#[path = "parser_test.rs"]
mod tests;
